### PR TITLE
Simulate BWC tests that start with the current version

### PR DESCRIPTION
**Pull request by alpar-t**
This changes does the following:
- version collections  now:
   - returns a fully qualified Gradle project for unreleased versions 
   - the current  version is not filtered out  when returning compatible versions 
   - we return `:distribution` for the current version 
- `:distribution` now emulates the configurations of a project that builds an unreleased version so   bwc tests can use it. 

This change causes the current version to be included as an "old" version in bwc tests, such that bwc tests will now test a currentVersion to currentVersion upgrade. 

This has the advantage of testing the plain restarts, but also makes the part of the test meant to run against the old cluster also run against the current version, this is important because in time the current version actually becomes the old version and suddenly we have a lot of test code that  was never ran against it. This change makes sure that this is not the case and version bumps can be smooth.

### CODE REVIEW

**Comment by atorok** [comment]
I was too quick to submit this one as it has changes from #39059. Added the WIP label.

**Comment by mark-vieira** [comment]
> I was too quick to submit this one as it has changes from #39059. Added the WIP label.

Please ping when this is ready fro review again.

**Comment by atorok** [comment]
@mark-vieira @jasontedor @rjernst 

This is now ready for review. The last test failure is being addressed by @jkakavas it's a flaky test which is made deterministic by this change ( the test assumed that a master on current version will not be elected in `oneThirdUpgraded` ). 

I will not merge this before that fix is in, but I think we can start the review process.

**Comment by mark-vieira** [comment]
Can't provide a lot of insight here as I'm not completely aware of all the business rules with regards to versions. I wouldn't mind sitting down with someone to go over the differences between bugfix, maintenance, staged, etc as well as our definitions of wire-compatible vs index-compatible.

**Comment by jasontedor** [comment]
@mark-vieira I am happy to do this! Please ping me tomorrow?

File: distribution/build.gradle [hunk 1]
```diff
@@ -516,3 +516,16 @@ subprojects {
     }
   }
 }
+
+['archives:windows-zip','archives:oss-windows-zip',
```
Comment (rjernst) [inline distribution/build.gradle]: Can you please explain this block?

File: buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java [hunk 2]
```diff
@@ -186,7 +183,7 @@ public void forPreviousUnreleased(Consumer<UnreleasedVersionInfo> consumer) {
 
     private String getGradleProjectNameFor(Version version) {
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java]: While @jasontedor and I were walking through this together we found it really hard to reason about this logic. I later ready the top-level javadoc and that cleared a few things up but I think in general this method in particular could use some comments to guide the reader along. Can we add some comments to the logic in this method to better explain what is happening?

**Review by mark-vieira — CHANGES_REQUESTED** [review]
While reading the javadoc there was a bit of confusion, with regards to the following:

```<li>the unreleased <b>bugfix</b>, M.N.c (c &gt; 0) on the `M.b` branch</li>```

I believe the branch in question here should be `M.N` as it's never explained what `b` is. As I understand it, work on `6.6.2` should be done in the `6.6` branch, yes?

We make a similar mention to an undefined `b` later in the javadoc:

```
* E.x when M.N.c is released M.N.c+1 is added to the Version class mentioned above in all the following branches:
*  `M.b`, `M.x` and `master` so we can reliably assume that the leafs of the version tree are unreleased.
```

Again, I think we should refer to branch `M.N` rather than `M.b`.

**Comment by atorok** [comment]
> Again, I think we should refer to branch `M.N` rather than `M.b`.

Fixed all of those, good catch, this was an oversight, it really should have been `N` rather than `b`.

File: buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java]: renamed, makes much more sense.

Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java]: I added a few comments and a pointer to the class javadoc

**Review by mark-vieira — APPROVED** [review]

**Comment by atorok** [comment]
Will back-port build setup only, without enabling the tests to make it easier to back-port future   build changes.

**Pull request merged**